### PR TITLE
Log numbers from `performance` in `@guardian/libs`

### PR DESCRIPTION
**Pull request by sndrs**
## What are you changing?

- renames `teams` in `logger` to `subscriptions`
- adds a `perf` subscription
- logs measurements triggered by `startPerformanceMeasure` when subscribed to `perf`

## Why?

- so we can how long things take from the console

262729139-0328860e-7494-46c5-87c1-7762b70a5e6c

### CODE REVIEW

File: libs/@guardian/libs/README.md [hunk 1]
```diff
@@ -66,6 +66,10 @@ Selectively log team-specific messages to the console.
 
 Types relating to Ophan.
 
+### [Performance](./src/performance)
+
+API over `window.performance` and hooks into the data warehouse.
+
```
Comment (mxdvl) [inline libs/@guardian/libs/README.md:72]: 👏

File: libs/@guardian/libs/src/deprecated-exports.ts [hunk 2]
```diff
@@ -5,6 +5,10 @@
  */
 
 import { Pillar } from './format/Pillar';
+import { type Subscription } from './logger/@types/logger';
 
 /** @deprecated Use `Pillar` instead. */
 export const ArticlePillar = Pillar;
+
+/** @deprecated Use `Subscription` instead. */
+export type TeamName = Subscription;
```
Comment (mxdvl) [inline libs/@guardian/libs/src/deprecated-exports.ts:14]: Nice one… would be nice to ensure this gets removed in the next major version.

File: libs/@guardian/libs/src/logger/log.ts [hunk 3]
```diff
@@ -2,45 +2,49 @@ import { isString } from '../isString/isString';
...
+	const subscriptions: string[] = getSubscriptions();
+	if (!subscriptions.includes(subscription)) subscriptions.push(subscription);
+	storage.local.set(STORAGE_KEY, subscriptions.join(','));
+	log(subscription, '🔔 Subscribed, hello!');
 };
 
 /**
- * Unsubscribe to a team’s log
- * @param team the team’s unique ID
+ * Unsubscribe to a subscription
```
Comment (mxdvl) [inline libs/@guardian/libs/src/logger/log.ts:41]: Keeping things symmetrical:

```suggestion
 * Unsubscribe from a log
```

File: libs/@guardian/libs/src/logger/@types/logger.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
...
+import type { commonStyle, subscriptionStyles } from '../subscriptions';
 
-export type Teams<K extends string> = Record<K, Record<string, string>>;
-export type LogCall = (team: TeamName, ...args: unknown[]) => void;
-export type TeamSubscription = (arg: TeamName) => void;
-export type TeamStyle = TeamName | keyof typeof commonStyle;
-export type TeamName = keyof typeof teamStyles;
+export type Subscriptions<K extends string> = Record<K, Record<string, string>>;
+export type LogCall = (subscription: Subscription, ...args: unknown[]) => void;
+export type SubscriptionStyle = (arg: Subscription) => void;
```
Comment (mxdvl) [inline libs/@guardian/libs/src/logger/@types/logger.ts:5]: I’m not sure this is aptly named… is this the method that allows you to subscribe and unsubscribe?

```suggestion
export type Subscribe = (arg: Subscription) => void;
```

File: libs/@guardian/libs/src/logger/log.ts [hunk 5]
```diff
@@ -51,17 +55,40 @@ if (typeof window !== 'undefined') {
...
+	if (!getSubscriptions().includes(subscription)) return;
+	if (subscription === 'perf') {
+		const [measurement, ...rest] = args;
+		const styles = [
+			messageStyle('common'),
+			'',
+			messageStyle(subscription),
+			// .replace(/background/g, 'temp')
+			// .replace(/color/g, 'background')
+			// .replace(/temp/g, 'color')
```
Comment (mxdvl) [inline libs/@guardian/libs/src/logger/log.ts:81]: 👀

File: libs/@guardian/libs/src/logger/logger.test.ts [hunk 6]
```diff
@@ -109,7 +109,7 @@ describe('Add and remove teams', () => {
 });
 
 describe('Team-based logging', () => {
-	const teams: TeamName[] = ['cmp', 'commercial', 'dotcom'];
+	const teams: Subscription[] = ['cmp', 'commercial', 'dotcom'];
```
Comment (mxdvl) [inline libs/@guardian/libs/src/logger/logger.test.ts:112]:
```suggestion
	const teams = ['cmp', 'commercial', 'dotcom'] satisfies Subscription[];
```

File: libs/@guardian/libs/src/logger/subscriptions.ts [hunk 7]
```diff
@@ -47,9 +47,13 @@ const teamStyles = {
 		background: '#C74600',
 		font: '#FEF9F5',
 	},
+	perf: {
+		background: '#FFD700',
+		font: '#000000',
+	},
 } as const;
```
Comment (mxdvl) [inline libs/@guardian/libs/src/logger/subscriptions.ts:54]: What about a little `satisfies`, here?

```suggestion
} as const satisfies Record<string, { background: string; font: string}>;
```

**Review by mxdvl — COMMENTED** [review]
Minor things outstanding, but otherwise a nice change!

File: libs/@guardian/libs/src/logger/@types/logger.ts (hunk 4 above)
Comment (sndrs) [reply libs/@guardian/libs/src/logger/@types/logger.ts:5]: oh yeah looks liek it - it's really hard to know with the types all on their own and now descriptions. it's a confusing pattern

File: libs/@guardian/libs/src/performance/startPerformanceMeasure.ts [hunk 8]
```diff
@@ -47,16 +52,20 @@ export const startPerformanceMeasure = (
 					duration: performance.now() - start,
 				};
 
-				return Math.ceil(duration);
+				const formattedDuration = Math.ceil(duration);
+
+				logPerf(measureName, formattedDuration);
```
Comment (mxdvl) [inline libs/@guardian/libs/src/performance/startPerformanceMeasure.ts:57]: “Formatted” or “rounded up to the nearest millisecond”?

**Review by mxdvl — APPROVED** [review]

File: libs/@guardian/libs/src/performance/startPerformanceMeasure.ts (hunk 8 above)
Comment (sndrs) [reply libs/@guardian/libs/src/performance/startPerformanceMeasure.ts:57]: i'd say "rounded up to the nearest millisecond" is how it's been formatted

File: libs/@guardian/libs/src/logger/logger.test.ts (hunk 6 above)
Comment (sndrs) [reply libs/@guardian/libs/src/logger/logger.test.ts:112]: what's the benefit of this?

Comment (mxdvl) [reply libs/@guardian/libs/src/logger/logger.test.ts:112]: It narrows the type to only these specific three subscriptions, which actually brings little value in a test.

**Pull request merged**